src/rrdp-iostat/rrdp_iostat.ml

```diff
@@ -320,8 +318,78 @@ module Blktap3_stats_wrapper = struct
       low_mem_mode		= false;
     }
 
+  let shm_devices_dir = "/dev/shm"
+
+  let m = Mutex.create ()
+  let cache : ((int * int), int option) Hashtbl.t = Hashtbl.create 100
+
+  let get_pids () =
+    let pid_from_xs (domid, devid) =
+      try
+        let result = with_xs (fun xs ->
+            let path = Printf.sprintf "/local/domain/0/backend/vbd3/%d/%d/kthread-pid" domid devid in
+            Some (int_of_string (xs.Xs.read path))) in
+        result
+      with _ ->
+        None (* Most likely the pid hasn't been written yet *)
+    in
+
+    Mutex.execute m (fun () ->
+        Hashtbl.fold (fun (domid, devid) pid_opt acc ->
+            match pid_opt with
+            | Some pid -> ((domid, devid), pid)::acc
+            | None ->
+              match pid_from_xs (domid, devid) with
+              | Some pid ->
+                Hashtbl.replace cache (domid, devid) (Some pid);
+                ((domid, devid), pid)::acc
+              | None -> acc) cache [])
+
+  let inotify_thread () =
+    let domdev_of_file f =
+      let domid, devid = Scanf.sscanf f "vbd3-%d-%d" (fun id devid -> (id, devid)) in
+      (domid, devid) in
+
+    let operate f fn =
+      let (domid, devid) = domdev_of_file f in
+      Mutex.execute m (fun () -> fn (domid, devid)) in
+
+    let add_file f = operate f (fun k -> D.debug "add_file: (%d,%d)" (fst k) (snd k); Hashtbl.replace cache k None) in
+    let remove_file f = operate f (fun k -> D.debug "remove_file: (%d,%d)" (fst k) (snd k); Hashtbl.remove cache k) in
+
+    let pred path = Xstringext.String.startswith "vbd3-" path in
+
+    let initialise () =
+      Mutex.execute m (fun () -> Hashtbl.clear cache);
+      let shm_dirs = Array.to_list (Sys.readdir shm_devices_dir) in
+      let shm_vbds = List.filter pred shm_dirs in
+      List.iter add_file shm_vbds;
+      D.debug "Populated %d cache entries" (List.length shm_vbds)
+    in
+
+    let inotify = Inotify.create () in
+    let _ = Inotify.add_watch inotify "/dev/shm/" Inotify.[S_Create; S_Delete] in
+    initialise ();
+
+    let rec loop () =
+      try
+        let evs = Inotify.read inotify in
+        List.iter (fun (w, kind_list, _, path) ->
+            let is_create = List.mem Inotify.Create kind_list in
```

Comment:
Is `kind_list` an ordered list of events? If so, we'd be interested in the latest one only.

Reply:
It seems to come from a bitmask... so: no.

Reply:
The important thing seems to be iterating through `evs` in order.